### PR TITLE
feat(pkg/catalog):  gateway is configured for SSL passthrough.

### PR DESCRIPTION
**Pull request by steeling**
The gateway is configured for SSL passthrough, and discovers every service in the local cluster.

Signed-off-by: Sean Teeling <seanteeling@microsoft.com>

**Affected area**:
| Functional Area            |     |
| -------------------------- | --- |
| New Functionality          | [ ] |
| Documentation              | [ ] |
| Install                    | [ ] |
| CLI Tool                   | [ ] |
| Certificate Management     | [ ] |
| Control Plane              | [ ] |
| Ingress                    | [ ] |
| Egress                     | [ ] |
| Networking                 | [ ] |
| Observability              | [ ] |
| SMI Policy                 | [ ] |
| Sidecar Injection          | [ ] |
| Security                   | [ ] |
| Upgrade                    | [ ] |
| Tests                      | [ ] |
| CI System                  | [ ] |
| Demo                       | [ ] |
| Performance                | [ ] |
| Other                      | [ ] |

Please answer the following questions with yes/no.

1. Does this change contain code from or inspired by another project?
    -   Did you notify the maintainers and provide attribution?

No

1. Is this a breaking change?
No.

### CODE REVIEW

File: pkg/catalog/outbound_traffic_policies.go [hunk 1]
```diff
@@ -135,6 +150,36 @@ func (mc *MeshCatalog) ListAllowedOutboundServicesForIdentity(serviceIdentity id
...
+			continue
+		}
+
+		weightedCluster := getDefaultWeightedClusterForService(destService)
+		policy := trafficpolicy.NewOutboundTrafficPolicy(destService.FQDN(), hostnames)
+		if err := policy.AddRoute(trafficpolicy.WildCardRouteMatch, weightedCluster); err != nil {
+			log.Error().Err(err).Msgf("Error adding route to outbound policy in permissive mode for destination %s(%s)", destService.Name, destService.Namespace)
+			continue
+		}
+		outPolicies = append(outPolicies, policy)
```
Comment (shashankram) [inline pkg/catalog/outbound_traffic_policies.go:178]: Why is an RDS route required for SNI passthrough clusters?

File: pkg/envoy/cds/cluster.go [hunk 2]
```diff
@@ -25,6 +25,22 @@ const (
...
+
+	return &xds_cluster.Cluster{
+		Name:                          upstreamSvc.String(),
+		ConnectTimeout:                ptypes.DurationProto(clusterConnectTimeout),
+		TypedExtensionProtocolOptions: HTTP2ProtocolOptions,
+		ClusterDiscoveryType:          &xds_cluster.Cluster_Type{Type: xds_cluster.Cluster_EDS},
+		EdsClusterConfig:              &xds_cluster.Cluster_EdsClusterConfig{EdsConfig: envoy.GetADSConfigSource()},
+		LbPolicy:                      xds_cluster.Cluster_ROUND_ROBIN,
+	}, nil
+}
```
Comment (shashankram) [inline pkg/envoy/cds/cluster.go:42]: We have been referring to passthrough clusters as clusters that use original destination type cluster discovery (ex. with global egress).  Can you not use `getUpstreamServiceCluster()` for your use case?

File: pkg/envoy/cds/response.go [hunk 3]
```diff
@@ -34,6 +29,33 @@ func NewResponse(meshCatalog catalog.MeshCataloger, proxy *envoy.Proxy, _ *xds_d
...
+			cluster, err := getPassthroughCluster(dstService)
+			if err != nil {
+				log.Error().Err(err).Msgf("Failed to construct service cluster for service %s for proxy with XDS Certificate SerialNumber=%s on Pod with UID=%s",
+					dstService.Name, proxy.GetCertificateSerialNumber(), proxy.String())
+				return nil, err
+			}
+
+			clusters = append(clusters, cluster)
+		}
+		return makeResources(proxy, clusters), nil
```
Comment (shashankram) [inline pkg/envoy/cds/response.go:56]: Should leverage the `getUpstreamServiceCluster()` here if cluster discovery is happening through EDS.

**Review by shashankram — CHANGES_REQUESTED** [review]
Hey @steeling, left a few comments that I'd like to be addressed. Would also prefer breaking up the changes as follows:
1. Introduce catalog API + tests
2. Introduce CDS integration + tests
3. Introduce LDS intergration + tests

File: pkg/envoy/cds/cluster.go (hunk 2 above)
Comment (steeling) [reply pkg/envoy/cds/cluster.go]: `getUpstreamServiceCluster`, sets the TLS context, which we don't want to do for SSL Passthrough.

Renamed to getSSLPasshthrough, but open to other suggestions.

In permissive mode, we also don't necessarily want to use the ORIGINAL_DST, since that will look for a cluster with a name where we don't have a DNS entry (ie: with the .global suffix)

File: pkg/envoy/cds/response.go (hunk 3 above)
Comment (steeling) [reply pkg/envoy/cds/response.go]: same commment as above

File: pkg/catalog/identity.go [hunk 4]
```diff
@@ -0,0 +1,14 @@
+package catalog
```
[the remark this answers is not in the document]
Comment (steeling) [reply pkg/catalog/identity.go]: done!

File: pkg/catalog/outbound_traffic_policies.go (hunk 1 above)
Comment (steeling) [reply pkg/catalog/outbound_traffic_policies.go]: oops, that was an old change, removed now.

File: pkg/envoy/cds/cluster.go (hunk 2 above)
Comment (shashankram) [reply pkg/envoy/cds/cluster.go]: Could you instead reuse this function and pass a bool for `useTLS bool`?

Comment (steeling) [reply pkg/envoy/cds/cluster.go]: added variadic options, IMO the other way is more clean though.

File: pkg/envoy/cds/cluster.go [hunk 5]
```diff
@@ -25,14 +25,30 @@ const (
...
+
 // getUpstreamServiceCluster returns an Envoy Cluster corresponding to the given upstream service
 // Note: ServiceIdentity must be in the format "name.namespace" [https://github.com/openservicemesh/osm/issues/3188]
-func getUpstreamServiceCluster(downstreamIdentity identity.ServiceIdentity, upstreamSvc service.MeshService, cfg configurator.Configurator) (*xds_cluster.Cluster, error) {
-	clusterName := upstreamSvc.String()
-	marshalledUpstreamTLSContext, err := ptypes.MarshalAny(
-		envoy.GetUpstreamTLSContext(downstreamIdentity, upstreamSvc))
-	if err != nil {
-		return nil, err
+func getUpstreamServiceCluster(downstreamIdentity identity.ServiceIdentity, upstreamSvc service.MeshService, cfg configurator.Configurator, opts ...Option) (*xds_cluster.Cluster, error) {
```
Comment (shashankram) [inline pkg/envoy/cds/cluster.go:48]: Hey @steeling, I didn't mean to suggest using variadic param here, but rather an additional param to use mTLS.

The goal here is to have a generic way to build clusters for different use cases, ex.
```
getUpstreamServiceCluster(clusterConfig []someStruct) (*xds_cluster.Cluster, error)
```

Since this is outside the scope of your PR, I suggested to simply pass a bool to not initialize the upstream TLS context when using SNI passthrough.

**Review by shashankram — COMMENTED** [review]
Hey @steeling my original request to split up the PR into individual units would still be beneficial to reviews and ensure a faster merge cycle.

File: pkg/envoy/cds/cluster.go (hunk 5 above)
Comment (steeling) [reply pkg/envoy/cds/cluster.go:48]: Sure, but we also need to disable permissive discovery. So there needs to be an additional bool option, at which variation Params would be more clean than tacking on multiple bool options
